vega_sim/scenario/common/agents.py

```diff
@@ -1125,6 +1129,90 @@ def step(self, vega_state: VegaState):
                 key_name=self.key_name,
             )
 
+    def _scale_orders(self, buy_shape: List[MMOrder], sell_shape: List[MMOrder]):
+
+        est_midprice = 0.5 * (buy_shape[0].price + sell_shape[0].price)
+
+        instantaneous_liquidity = min(
+            [
+                self._calculate_liquidity(
+                    side=vega_protos.SIDE_BUY,
+                    orders=buy_shape,
+                    midprice=est_midprice,
+                ),
+                self._calculate_liquidity(
+                    side=vega_protos.SIDE_SELL,
+                    orders=sell_shape,
+                    midprice=est_midprice,
+                ),
+            ]
+        )
+
+        # Compute the scaling factor
+        stake_to_ccy_siskas = self.vega.get_network_parameter(
+            "market.liquidity.stakeToCcySiskas",
+            "float",
+        )
+        scaling_factor = (
+            self.commitment_amount * stake_to_ccy_siskas
+        ) / instantaneous_liquidity
+
+        # Scale the shapes
+        scaled_buy_shape = [
+            MMOrder(order.size * scaling_factor, order.price) for order in buy_shape
+        ]
+        scaled_sell_shape = [
+            MMOrder(order.size * scaling_factor, order.price) for order in sell_shape
+        ]
+
+        return scaled_buy_shape, scaled_sell_shape
+
+    def _calculate_liquidity(
+        self, side: vega_protos.side, orders: List[MMOrder], midprice: float
+    ) -> float:
+
+        market_info = self.vega.market_info(market_id=self.market_id)
+        tau = market_info.tradable_instrument.log_normal_risk_model.tau
+        mu = market_info.tradable_instrument.log_normal_risk_model.params.mu
+        sigma = market_info.tradable_instrument.log_normal_risk_model.params.sigma
+
+        tau_scaling = self.vega.get_network_parameter(
+            "market.liquidity.probabilityOfTrading.tau.scaling",
+            "float",
+        )
+        min_probability_of_trading = self.vega.get_network_parameter(
+            "market.liquidity.minimum.probabilityOfTrading.lpOrders",
+            "float",
+        )
+
+        (min_valid_price, max_valid_price) = self.vega.price_bounds(
+            market_id=self.market_id
+        )
+
+        provided_liquidity = 0
+
+        for vol, price in orders:
+
+            p = max(
+                [
+                    min_probability_of_trading,
+                    probability_of_trading(
+                        tau=tau * tau_scaling,
+                        mu=mu,
+                        sigma=sigma,
+                        lower_bound=min_valid_price,
+                        upper_bound=max_valid_price,
+                        best_price=midprice,
+                        price=price,
+                        side=side,
+                    ),
+                ]
+            )
+
+            provided_liquidity += vol * price * p
```

Comment:
Instinctively, this loop (and so ultimately the `probability_of_trading` function) can almost certainly be vectorised and run through numpy functions. Our lists probably aren't long enough that it's worth doing now but if you ever want to have a shot then give it a go 😄

Reply:
Thanks yeah I will definitely come back to this at some point!